rust/src/dcerpc/dcerpc.rs

```diff
@@ -0,0 +1,2600 @@
+            self.buffer_ts.clear();
+        }
+        return 0;
+    }
+
+    pub fn handle_gap_tc(&mut self) -> u8 {
+        if self.buffer_tc.len() > 0 {
+            self.buffer_tc.clear();
+        }
+        return 0;
+    }
+
+    pub fn clean_buffer(&mut self, direction: u8) {
+        match direction {
+            core::STREAM_TOSERVER => {
+                self.buffer_ts.clear();
+            }
+            _ => {
+                self.buffer_tc.clear();
+            }
+        }
+        self.bytes_consumed = 0;
+    }
+
+    pub fn extend_buffer(&mut self, buffer: &[u8], direction: u8) {
+        match direction {
+            core::STREAM_TOSERVER => {
+                self.buffer_ts.extend_from_slice(buffer);
+            }
+            _ => {
+                self.buffer_tc.extend_from_slice(buffer);
+            }
+        }
+        self.data_needed_for_dir = direction;
+    }
+
+    pub fn reset_direction(&mut self, direction: u8) {
+        if direction == core::STREAM_TOSERVER {
+            self.data_needed_for_dir = core::STREAM_TOCLIENT;
+        } else {
+            self.data_needed_for_dir = core::STREAM_TOSERVER;
+        }
+    }
+
+    pub fn process_header(&mut self, input: &[u8]) -> i32 {
+        match parser::parse_dcerpc_header(input) {
+            Ok((leftover_bytes, header)) => {
+                if header.rpc_vers != 5
+                    || (header.rpc_vers_minor != 0 && header.rpc_vers_minor != 1)
+                {
+                    SCLogDebug!(
+                        "DCERPC Header did not validate. Major version: {:?} Minor version: {:?}",
+                        header.rpc_vers,
+                        header.rpc_vers_minor
+                    );
+                    return -1;
+                }
+                self.header = Some(header);
+                (input.len() - leftover_bytes.len()) as i32
+            }
+            Err(nom::Err::Incomplete(_)) => {
+                // Insufficient data.
+                SCLogDebug!("Insufficient data while parsing DCERPC header");
+                -1
+            }
+            Err(_) => {
+                // Error, probably malformed data.
+                SCLogDebug!("An error occurred while parsing DCERPC header");
+                -1
+            }
+        }
+    }
+
+    pub fn handle_bindctxitem(&mut self, input: &[u8], uuid_internal_id: u16) -> i32 {
+        let endianness = self.get_endianness();
+        match parser::parse_bindctx_item(input, endianness) {
+            Ok((leftover_bytes, ctxitem)) => {
+                let mut uuidentry = DCERPCUuidEntry::new();
+                uuidentry.uuid = ctxitem.uuid;
+                uuidentry.internal_id = uuid_internal_id;
+                uuidentry.ctxid = ctxitem.ctxid;
+                uuidentry.version = ctxitem.version;
+                uuidentry.versionminor = ctxitem.versionminor;
+                let pfcflags = match self.get_hdr_pfcflags() {
```

Comment:
Not tested, could this be `let pfcflags = self.get_hdr_pfcflags().unwrap_or(0);`?